File: java/Enemy.java
import java.lang.Math;

public class Enemy extends BoardObject {

	private int maxSpeed;
	private int speed;

	public Enemy(int x, int y, int speed) {
		super(x, y, "X");
		this.maxSpeed = speed;
		this.speed = speed;
	}

	public void up() {
		if (y>1)
			y -= 1;
	}

	public void down() {
		if (y<Board.HEIGHT-2)
			y += 1;
	}

	public void right() {
		if (x<Board.WIDTH-2)
			x += 1;
	}

	public void left() {
		if (x>1)
			x -= 1;
	}

	public void move(Player player) {
		if (Math.abs(player.x()-x) < Math.abs(player.y()-y)) {
			if (player.y() < y)
				up();
			else
				down();
		} else {
			if (player.x() > x)
				right();
			else
				left();
		}
	}

	public boolean isCollidingPlayer(Player player) {
		if (Math.abs(player.x()-x)<=1 && Math.abs(player.y()-y)<=1)
			return true;
		else 
			return false;
	}

	public boolean process(Player player) {
		speed -= 1;
		if (speed == 0) {
			speed = maxSpeed;
			move(player);
		}
		return isCollidingPlayer(player);	
	}
}

File: rust/racism_exe/src/enemy.rs
use crate::{miscellaneous::Entity, player::Player};

#[derive(Debug, Clone)]
pub struct Enemy {
	pub x: u32,
	pub y: u32,
	can_move: bool,
	pub char: String,
}

impl Enemy {
	pub fn new(width: u32, height: u32) -> Self {
		Self {
			x: width-2,
			y: height-2,
			can_move: false,
			char: "☺".to_string(),
		}
	}

	pub fn is_colliding_player(self, player: Player) -> bool {
		let dist_x = (self.x as i32 - player.x as i32).abs();
		let dist_y = (self.y as i32 - player.y as i32).abs();
		dist_x <= 1 && dist_y <= 1 
	}

	pub fn process(&mut self, player: Player, width: u32, height: u32) {
		if self.can_move {
			self.can_move = false;
			let dist_x = (self.x as i32 - player.x as i32).abs();
			let dist_y = (self.y as i32 - player.y as i32).abs();
			if dist_x >= dist_y {
				if self.x > player.x {
					self.left();
				} else {
					self.right(width);
				}
			} else {
				if self.y > player.y {
					self.up();
				} else {
					self.down(height);
				}
			}
		} else {
			self.can_move = true;
		}
	}
}

impl Entity for Enemy {
    fn up(&mut self) {
        if self.y > 1 {
        	self.y -= 1;
        }
    }

    fn down(&mut self, height: u32) {
        if self.y < height - 2 {
        	self.y += 1;
        }
    }

    fn right(&mut self, width: u32) {
    	if self.x < width - 2 {
    		self.x += 1;
    	}
    }

    fn left(&mut self) {
        if self.x > 1 {
        	self.x -= 1;
        }
    }
}